cpp/src/tsne/distances.cuh

```diff
@@ -174,5 +182,22 @@ void symmetrize_perplexity(float* P,
     indices, P, n, k, COO_Matrix, stream);
 }
 
+template <typename value_t>
+value_t compute_grad_norm(const raft::handle_t& handle,
+                          value_t* old_forces_device,
+                          std::size_t n_pts)
+{
+  value_t grad_norm = thrust::transform_reduce(handle.get_thrust_policy(),
+                                               old_forces_device,
+                                               old_forces_device + n_pts,
+                                               FunctionalSquare(),
+                                               0.0,
+                                               thrust::plus<value_t>());
```

Comment:
Besides adding early stopping to B-H, this PR actually corrects the "grad_norm" that the FFT implementation had (identical bug in CannyLab's: https://github.com/CannyLab/tsne-cuda/issues/90). The values correlate well, but are not truly the same, and more importantly are ~two orders of magnitude different, so most old `min_grad_norm` values won't work anymore, nor will the default `min_grad_norm` of 1e-7.

![image](https://user-images.githubusercontent.com/469365/144786042-0c9398ce-3c51-4524-8ea7-06f4a9c692fb.png)